app/policies/member_policy.rb

```diff
@@ -0,0 +1,19 @@
+# frozen_string_literal: true
+
+# Base policy for member authorization
+class MemberPolicy < ApplicationPolicy
+  scope_for :relation do |relation, options|
+    namespace = options[:namespace]
+
+    namespace_ids = if namespace.project_namespace?
+                      [namespace.id] + namespace.parent&.self_and_ancestor_ids
+                    else
+                      namespace.self_and_ancestor_ids
+                    end
+
+    relation.where(id: Member.includes(namespace: [:route]).where(namespace_id: namespace_ids)
```

Comment:
**suggestion (blocking):** This morning I was able to come up with a much better query than I had previously suggested. The following query does not use `uniq` or `pluck` which means that it can all happen in sql at once without having to perform the inner query before executing the outer query.

```ruby
relation.where(id: Member.joins(namespace: [:route]).where(namespace_id: namespace_ids)
                                                        .order(user_id: :asc, 'routes.path': :desc)
                                                        .select('DISTINCT ON (user_id) members.id'))
```